core/src/main/java/io/grpc/internal/OverrideAuthorityNameResolverFactory.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2017, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.Attributes;
+import io.grpc.NameResolver;
+import java.net.URI;
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper class that overrides the authority of a NameResolver, while preserving all other
+ * functionality.
+ */
+final class OverrideAuthorityNameResolverFactory extends NameResolver.Factory {
+  private final NameResolver.Factory delegate;
+  private final String authorityOverride;
+
+  /**
+   * Constructor for the {@link NameResolver.Factory}
+   *
+   * @param delegate The actual underlying factory that will produce the a {@link NameResolver}
+   * @param authorityOverride The authority that will be returned by {@link
+   *   NameResolver#getServiceAuthority()}
+   */
+  OverrideAuthorityNameResolverFactory(NameResolver.Factory delegate,
+      String authorityOverride) {
+    this.delegate = delegate;
+    this.authorityOverride = authorityOverride;
+  }
+
+  @Nullable
+  @Override
+  public NameResolver newNameResolver(URI targetUri, Attributes params) {
+    final NameResolver resolver = delegate.newNameResolver(targetUri, params);
+    // Do not wrap null values. We do not want to impede error signaling.
+    if (resolver == null) {
+      return null;
+    }
+    return new NameResolver() {
+      @Override
```

Comment:
This is pre-existing code, but you're right. That's a bug. Let's fix that separately to avoid hidden behavior changes in this PR. Filed #3061